README
Perl tools for www.rajce.idnes.cz.


INSTALL
=======

sudo cpan WebService::Rajce


USAGE AND INFO
==============

man rajce-get
man rajce-put


SEE ALSO
========
Rajce API - http://goo.gl/GNFSk
CPAN - http://search.cpan.org/perldoc?WebService::Rajce
Source - https://github.com/petrkle/rajce


Changes
{{$NEXT}}

1.202830  2020-10-09 15:23:58+02:00 Europe/Prague
	Adaptation to changes on web rajce.net, remove --sort option

1.190840  2019-03-25 15:40:36+01:00 Europe/Prague
	Adaptation to changes on web rajce.net - thanks to Jan Bubik.

1.180380  2018-02-07 13:25:51+01:00 Europe/Prague
	Restore http support - thanks Petr Písař

1.180370  2018-02-06 14:46:28+01:00 Europe/Prague
	https support

1.152450  2015-09-02 18:17:28+02:00 Europe/Prague
	Add long command line switches

1.152420  2015-08-30 13:41:02CEST+0200 Europe/Prague
	Add option -s to rajce-get: Prefix output file names with numbers
	to keep ordering in album. Thanks to Petr Spacek.

1.130930  2013-04-03 22:34:02 Europe/Prague
	Move to Dist::Zila.

0.08		Oct 11 2012
	Add -k switch - keep exif data.

0.07		Jan 03 2012
	Add -V switch - show version.

0.06		Dec 22 2011
	Print help to STDOUT.
	Link to repository in META.
		
0.05		Dec 21 2011
	Fix -h switch in rajce-put.

0.04		Dec 19 2011
	Proxy support, -h switch.

0.03		Dec 15 2011
	Debuging and verbose output.

0.02		Dec 13 2011
	Add all required modules to build script.

0.01    Dec 11 2011
	First version.


WebService-Rajce/Build.PL
use strict;
use Module::Build;

Module::Build
  ->new( module_name => "WebService::Rajce",
         license     => 'perl',
         bin_files => [ 'bin/rajce-get', 'bin/rajce-put$' ],
				 dist_author         => 'Petr Kletecka <pek@cpan.org>',
				 dist_version_from   => 'lib/WebService/Rajce.pm',

				 configure_requires => {
					 'Module::Build' 		=> 0.38 
				 },
         build_requires => {
             'Test::More'     => 0,
         },
         requires    => {
             'WWW::Mechanize' => 1.50,
             'File::Path'	    => 2.06_05,
             'File::Basename'	=> 1.2,
             'Getopt::Std'		=> 0,
             'Net::Netrc'			=> 1.22,
             'XML::Simple'		=> 0,
             'Encode'					=> 0,
             'Image::Magick'	=> 0,
             'Digest::MD5'		=> 0,
						 'Pod::Usage'			=> 0,
						 'Carp'						=> 0,
         },
         create_makefile_pl => 'traditional',
  				meta_merge     => {
				    resources => {
				      repository  => 'https://github.com/petrkle/rajce/'
    				}
					},

       )
  ->create_build_script;


WebService-Rajce/Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.3800
use ExtUtils::MakeMaker;
WriteMakefile
(
  'PL_FILES' => {},
  'INSTALLDIRS' => 'site',
  'NAME' => 'WebService::Rajce',
  'EXE_FILES' => [
                   'bin/rajce-get',
                   'bin/rajce-put'
                 ],
  'VERSION_FROM' => 'lib/WebService/Rajce.pm',
  'PREREQ_PM' => {
                   'Pod::Usage' => 0,
                   'Getopt::Std' => 0,
                   'WWW::Mechanize' => '1.5',
                   'XML::Simple' => 0,
                   'Net::Netrc' => '1.22',
                   'Image::Magick' => 0,
                   'Encode' => 0,
                   'Carp' => 0,
                   'Test::More' => 0,
                   'Digest::MD5' => 0,
                   'File::Path' => '2.0605',
                   'File::Basename' => '1.2'
                 }
)
;


lib/WebService/Rajce.pm
package WebService::Rajce;
# ABSTRACT: Perl module for rajce.net web API.

use 5.006;
use strict;
use warnings;

use WWW::Mechanize;
use XML::Simple;
use Digest::MD5 qw(md5_hex);
use Encode;
use Image::Magick;
use Carp;

use vars qw($VERSION @ISA @EXPORT @EXPORT_OK);

require Exporter;


our @ISA = qw(Exporter AutoLoader);
our @EXPORT = qw();

=head1 SYNOPSIS

	use WebService::Rajce;
	my $rajce = new WebService::Rajce;
	$rajce->login($mail,$password);
	my $album = $rajce->create_album('Title','Description');
	$rajce->add_photo('/path/to/file.jpg',$album)

=method my $rajce = new WebService::Rajce;

Create new object instance.

The C<debug> param turns on debug mode.

The C<keep_exif> param keep EXIF information in uploaded image.
=cut
sub new {
	my $class = shift;
  my %passed_parms = @_;
	my $self  = {};
	$self->{API} = 'https://www.rajce.idnes.cz/liveAPI/index.php';
	$self->{XML} = '<?xml version="1.0" encoding="utf-8"?>';
	$self->{DEBUG} = $passed_parms{'debug'};
	$self->{KEEP_EXIF} = $passed_parms{'keep_exif'};
	$self->{BOT} = WWW::Mechanize->new(autocheck => 1, agent => 'github.com/petrkle/rajce - '.$VERSION);
	$self->{BOT}->env_proxy();
	$self->{BOT}->add_header('Accept-Encoding'=>'text/html');
	$self->{BOT}->add_header('Accept-Charset'=>'utf-8');
	$self->{BOT}->add_header('Accept-Language'=>'cs');
	$self->{BOT}->cookie_jar(HTTP::Cookies->new());

	$self->{ERRORS} = {
	'1' => 'Unknown error.',
	'2' => 'Invalid command.',
	'3' => 'Invalid login or password.',
	'4' => 'Bad login token.',
	'5' => 'Unknown or repeating column {colName}.',
	'6' => 'Not correct albumID.',
	'7' => 'Album not exist or logged user is not owner.',
	'8' => 'Bad album token.',
	'9' => 'Albumn cant have empty title.',
	'10' => 'Failed to create new album. (hard to say why ... probably an error on the server side).',
	'11' => 'Album not exist.',
	'12' => 'Non existing application.',
	'13' => 'Wrong application key.',
	'14' => 'File is not attached.',
	'15' => 'Already there is a newer version {version}.',
	'16' => 'Error when saving a file.',
	'17' => 'Illegal file extension {extension}.',
	'18' => 'Wrong version number of client.',
	'19' => 'No such object (target).',
	'20' => 'Missing name to protect the album.',
	'21' => 'Missing password to protect the album.',
	'22' => 'Error communication - arrived an empty file.',
	'23' => 'Some blocks of the video are missing.',
	'24' => 'User does not exist.',
	'25' => 'There is the correct userID or albumID.',
	'26' => 'Album not exist or is not belog to this user or isnt public.',
	'27' => 'Invalid clientVideoID.',
	'28' => 'Upload with a given number does not exist.',

	};

	bless($self, $class);
	return $self;
}

=method $rajce->_debug($message);

Show debugging message.
=cut
sub _debug{
	my ($self,$message) = @_;
	if($self->{DEBUG}){
		print encode("utf8",$message)."\n";
	}
}

=method $rajce->login($mail,$password);

Login to API.
=cut
sub login {
	my ($self,$mail,$password) = @_;

	my $login = {'request'=>{
			'command'=>['login'],
			'parameters'=>{
				'clientID'=>['Rajce.pm'],
				'currentVersion'=>[$VERSION],
				'lang'=>['cs_CZ'],
				'login'=>[$mail],
				'password'=>[md5_hex($password)]
			},
		}
	};

	my $xml = XMLout($login, KeepRoot => 1, XMLDecl => $self->{XML});
	$self->_debug($xml);
	my $sign = $self->{BOT}->post($self->{API}, {'data' => $xml});
	my $response = XMLin($sign->content());
	$self->_debug($sign->content());

	if($response->{errorCode}){
		confess($self->{ERRORS}{$response->{errorCode}});
	}

	$self->{sessionToken}=$response->{sessionToken};
	$self->{maxWidth}=$response->{maxWidth};
	$self->{maxHeight}=$response->{maxHeight};
	$self->{nick}=$response->{nick};
	return $response;
}

=method $rajce->list($userid);

Get list of albums.
NOTICE - list other users albums not implemented in API yet
=cut
sub list {
	my ($self,$userid) = @_;

	my $listalbums = {'request'=>{
			'command'=>['getAlbumList'],
			'parameters'=>{
				'token'=>[$self->{sessionToken}],
				'userID'=>[$userid],
			},
		}
	};

	my $xml = XMLout($listalbums, KeepRoot => 1, XMLDecl => $self->{XML});
	$self->_debug($xml);
	my $albums = $self->{BOT}->post($self->{API}, {'data' => $xml});
	$self->_debug($albums->content());

return XMLin($albums->content());
}

=method $rajce->photo_list($albumid);

Get list of images in album.
=cut
sub photo_list {
	my ($self,$albumid) = @_;

	my $photolist = {'request'=>{
			'command'=>['getPhotoList'],
			'parameters'=>{
				'token'=>[$self->{sessionToken}],
				'albumID'=>[$albumid],
				'columns'=>{
					'column'=>[
						'date',
						'name',
						'description',
						'url',
						'thumbUrl',
						'thumbUrlBest',
						'urlBase']
				}
			}
		}
	};

	my $xml = XMLout($photolist, KeepRoot => 1, XMLDecl => $self->{XML});
	$self->_debug($xml);
	my $photos = $self->{BOT}->post($self->{API}, {'data' => $xml});
	$self->_debug($photos->content());

return XMLin($photos->content());
}

=method $rajce->search_users($query,$skip,$limit);

Get list of users.
NOTICE - not implemented in API yet
=cut
sub search_users {
	my ($self,$query,$skip,$limit) = @_;

	my $users = {'request'=>{
			'command'=>['searchUsers'],
			'parameters'=>{
				'token'=>[$self->{sessionToken}],
				'query'=>[$query],
				'skip'=>[$skip],
				'limit'=>[$limit],
				'columns'=>{
					'column'=>[
					'fullName',
					'albumCount',
					'viewCount']
				}
			}
		}
	};

	my $xml = XMLout($users, KeepRoot => 1, XMLDecl => $self->{XML});
	$self->_debug($xml);
	my $result = $self->{BOT}->post($self->{API},	{'data' => $xml});
	$self->_debug($result->content());

return XMLin($result->content());
}

=method $rajce->get_url($target);

Get some URL from rajce.net
$target = 'user-profile' | 'email-notifications' | 'service-notifications' ;
=cut
sub get_url {
	my ($self,$target) = @_;

	my $geturl = {'request'=>{
			'command'=>['getUrl'],
			'parameters'=>{
				'token'=>[$self->{sessionToken}],
				'target'=>[$target],
			}
		}
	};

	my $xml = XMLout($geturl, KeepRoot => 1, XMLDecl => $self->{XML});
	$self->_debug($xml);
	my $result = $self->{BOT}->post($self->{API},	{'data' => $xml});
	my $response = XMLin($result->content());
	$self->_debug($response->content());

return $response->{url};
}


=method $rajce->search_albums($query,$skip,$limit);

Get list of users.
NOTICE - not implemented in API yet
=cut
sub search_albums {
	my ($self,$query,$skip,$limit) = @_;

	my $albums = {'request'=>{
			'command'=>['searchAlbums'],
			'parameters'=>{
				'token'=>[$self->{sessionToken}],
				'query'=>[$query],
				'skip'=>[$skip],
				'limit'=>[$limit],
				'columns'=>{
					'column'=>[
						'description',
						'shortenedDescription',
						'viewCount',
						'mediaCount',
						'createDate']
				}
			},
		}
	};

	my $xml = XMLout($albums, KeepRoot => 1, XMLDecl => $self->{XML});
	$self->_debug($xml);
	my $response = $self->{BOT}->post($self->{API},	{'data' => $xml});
	$self->_debug($response->content());

return XMLin($response->content());
}


=method $rajce->reg_url();

Get URL where is form for creating new account on rajce.net.
=cut
sub reg_url {
	my ($self) = @_;

	my $request = {'request'=>{
			'command'=>['getRegisterUrl']
		}
	};

	my $xml = XMLout($request, KeepRoot => 1, XMLDecl => $self->{XML});
	$self->_debug($xml);
	my $regurl = $self->{BOT}->post($self->{API},	{'data' => $xml});
	$self->_debug($regurl->content());

	my $response = XMLin($regurl->content());

	if($response->{errorCode}){
		confess($self->{ERRORS}{$response->{errorCode}});
	}

return $response->{url};
}

=method $rajce->recover_url();

Get URL where is form for recover forget password.
=cut
sub recover_url {
	my ($self) = @_;

	my $request = {'request'=>{
			'command'=>['getRecoverPasswordUrl']
		}
	};

	my $xml = XMLout($request, KeepRoot => 1, XMLDecl => $self->{XML});
	$self->_debug($xml);
	my $url = $self->{BOT}->post($self->{API}, {'data' => $xml});
	$self->_debug($url->content());

	my $response = XMLin($url->content());

	if($response->{errorCode}){
		confess($self->{ERRORS}{$response->{errorCode}});
	}

return $response->{url};
}

=method $rajce->create_album($title,$desc);

Create new album.
=cut
sub create_album {
	my ($self,$title,$desc) = @_;

	my $create = {'request'=>{
			'command'=>['createAlbum'],
			'parameters'=>{
				'token'=>[$self->{sessionToken}],
				'albumName'=>[decode("utf8",$title)],
				'albumDescription'=>[decode("utf8",$desc)],
				'albumVisible'=>[1],
			},
		}
	};

	my $xml = XMLout($create, KeepRoot => 1, XMLDecl => $self->{XML});
	$self->_debug($xml);
	my $album = $self->{BOT}->post($self->{API}, {'data' => $xml});
	$self->_debug($album->content());

	my $response = XMLin($album->content());
	if($response->{errorCode}){
		confess($self->{ERRORS}{$response->{errorCode}});
	}

return $response;
}

=method $rajce->_open_album($album);

Open album for adding pictures.
=cut
sub _open_album {
	my ($self,$album) = @_;

	my $request = {'request'=>{
			'command'=>['openAlbum'],
			'parameters'=>{
				'token'=>[$self->{sessionToken}],
				'albumID'=>[$album->{albumID}],
			},
		}
	};

	my $xml = XMLout($request, KeepRoot => 1, XMLDecl => $self->{XML});
	$self->_debug($xml);
	my $open = $self->{BOT}->post($self->{API},	{'data' => $xml});
	$self->_debug($open->content());

	my $response = XMLin($open->content());

	if($response->{errorCode}){
		confess($self->{ERRORS}{$response->{errorCode}});
	}

return $response;
}

=method $rajce->_close_album($album);

Close album after adding pictures.
=cut
sub _close_album {
	my ($self,$album) = @_;

	my $request = {'request'=>{
			'command'=>['closeAlbum'],
			'parameters'=>{
				'token'=>[$self->{sessionToken}],
				'albumToken'=>[$album->{albumToken}],
			},
		}
	};

	my $xml = XMLout($request, KeepRoot => 1, XMLDecl => $self->{XML});
	$self->_debug($xml);
	my $close = $self->{BOT}->post($self->{API}, {'data' => $xml});
	$self->_debug($close->content());

	my $response = XMLin($close->content());

	if($response->{errorCode}){
		confess($self->{ERRORS}{$response->{errorCode}});
	}

return $response;
}

=method $rajce->add_photo($filename,$album);

Add photo into gallery.
=cut
sub add_photo {
	my ($self,$filename,$album) = @_;

	my $thumbsize = "100x100";

	my $thumb = new Image::Magick;
	$thumb->Read($filename);
	$thumb->AutoOrient();
	$thumb->Resize(geometry=>"$thumbsize^");
	$thumb->Crop(gravity=>"Center",geometry=>"$thumbsize");
	$thumb->Strip();

	my $pic = new Image::Magick;
	$pic->Read($filename);
	$pic->AutoOrient();
	$pic->Resize(geometry=>"$self->{maxWidth}x$self->{maxHeight}>");

	if(!$self->{KEEP_EXIF}){
		$pic->Strip();
	}

	my ($width, $height) = $pic->Get('width','height');
	
	my $request = {'request'=>{
			'command'=>['addPhoto'],
			'parameters'=>{
				'token'=>[$self->{sessionToken}],
				'albumToken'=>[$album->{albumToken}],
				'width'=>[$width],
				'height'=>[$height],
			},
		}
	};

	$self->_open_album($album);

	my $xml = XMLout($request, KeepRoot => 1,	XMLDecl => $self->{XML});
	$self->_debug($xml);

	my $picture = $self->{BOT}->post($self->{API},
		{'data' => $xml,
			'thumb' => [undef,$filename,Content => $thumb->ImageToBlob()],
			'photo' => [undef,$filename,Content => $pic->ImageToBlob()]},
		Content_Type => 'form-data');


	$self->_debug($picture->content());
	my $response = XMLin($picture->content());


	if($response->{errorCode}){
		confess($self->{ERRORS}{$response->{errorCode}});
	}

	$self->_close_album($album);

return $response;
}

=method $rajce->get_albumurl($album);

Get URL of album.
=cut
sub get_albumurl {
	my ($self,$album) = @_;

	my $url = {'request'=>{
			'command'=>['getAlbumUrl'],
			'parameters'=>{
				'token'=>[$self->{sessionToken}],
				'albumToken'=>[$album->{albumToken}],
			}
		}
	};

	my $xml = XMLout($url, KeepRoot => 1, XMLDecl => $self->{XML});
	$self->_debug($xml);
	my $alb = $self->{BOT}->post($self->{API}, {'data' => $xml});
	$self->_debug($alb->content());
	my $response = XMLin($alb->content());

	if($response->{errorCode}){
		confess($self->{ERRORS}{$response->{errorCode}});
	}

return $response->{url};
}

1;
__END__

=head1 SEE ALSO

https://rajce.net/static/doc/LiveApi.html

=cut
1;
